openedx/core/djangoapps/user_api/management/commands/populate_retirement_states.py

```diff
@@ -0,0 +1,115 @@
+"""
+Migrates user preferences from one language code to another in batches. Dark lang preferences are not affected.
+"""
+from __future__ import print_function
+
+import copy
+import logging
+
+from django.conf import settings
+from django.core.management.base import BaseCommand, CommandError
+
+from openedx.core.djangoapps.user_api.models import RetirementState, UserRetirementStatus
+
+
+LOGGER = logging.getLogger(__name__)
+
+START_STATE = 'PENDING'
+END_STATES = ['ERRORED', 'ABORTED', 'COMPLETE']
+REQUIRED_STATES = copy.deepcopy(END_STATES)
+REQUIRED_STATES.insert(0, START_STATE)
+REQ_STR = ','.join(REQUIRED_STATES)
+
+
+class Command(BaseCommand):
+    """
+    Implementation of the populate command
+    """
+    help = 'Populates the RetirementState table with the states present in settings.'
+
+    def _validate_new_states(self, new_states):
+        """
+        Check settings for existence of states, required states
+        """
+        if not new_states:
+            raise CommandError('settings.RETIREMENT_STATES does not exist or is empty.')
+
+        if not set(REQUIRED_STATES).issubset(set(new_states)):
+            print(new_states)
+            raise CommandError('settings.RETIREMENT_STATES does not contain all required states ({})'.format(REQ_STR))
+
+        # Confirm that the start and end states are in the right places
+        if new_states.index(START_STATE) is not 0:
+            raise CommandError('{} must be the first state'.format(START_STATE))
+
+        num_end_states = len(END_STATES)
+
+        if new_states[-num_end_states:] != END_STATES:
+            raise CommandError('The last {} states must be these (in this order): '
+                               '{}'.format(num_end_states, END_STATES))
```

Comment:
hm, if the start and end states are always enforced, can't the settings just specify the ones in between?  Then we can append/prepend the required states internally without needing to do this validation.  Instead you could just check if the REQUIRED_STATES intersection with new_states is empty.

Reply:
We could, but I don't like the idea of having the states broken up any more than they already are (ie. when you look for them they could be coming from any of 5 repos, depending on the system you're looking at). I think it would be confusing to have a list of states in the place that's supposed to be definitive, with some others magically appearing in the database when they're imported. This also provides a full list that retirement drivers can access from one place in configuration instead of having to do that munging again in the driver implementation.